src/settings_gui.cpp

```diff
@@ -166,6 +167,171 @@ static const std::map<int, StringID> _volume_labels = {
+	EndContainer(),
+};
+
+class NWidgetSocialPlugins : public NWidgetVertical {
+public:
+	NWidgetSocialPlugins()
+	{
+		this->plugins = SocialIntegration::GetPlugins();
+
+		if (this->plugins.size() == 0) {
+			auto widget = MakeNWidgets(std::begin(_nested_social_plugins_none_widgets), std::end(_nested_social_plugins_none_widgets), nullptr);
+			this->Add(std::move(widget));
+		} else {
+			for (size_t i = 0; i < this->plugins.size(); i++) {
+				auto widget = MakeNWidgets(std::begin(_nested_social_plugins_widgets), std::end(_nested_social_plugins_widgets), nullptr);
+				this->Add(std::move(widget));
+			}
+		}
+
+		this->SetPIP(0, WidgetDimensions::unscaled.vsep_wide, 0);
+	}
+
+	void FillWidgetLookup(WidgetLookup &widget_lookup) override
+	{
+		widget_lookup[WID_GO_SOCIAL_PLUGINS] = this;
+		NWidgetVertical::FillWidgetLookup(widget_lookup);
+	}
+
+	void SetupSmallestSize(Window *w) override
+	{
+		this->current_index = -1;
+		NWidgetVertical::SetupSmallestSize(w);
+	}
+
+	/**
+	 * Find of all the plugins the one where the member is the widest (in pixels).
+	 *
+	 * @param member The member to check with.
+	 * @return The plugin that has the widest value (in pixels) for the given member.
+	 */
+	template <typename T>
+	std::string &GetWidestPlugin(T SocialIntegrationPlugin::*member) const
+	{
+		std::string *longest = &(this->plugins[0]->*member);
+		int longest_length = 0;
+
+		for (auto *plugin : this->plugins) {
+			int length = GetStringBoundingBox(plugin->*member).width;
+			if (length > longest_length) {
+				longest_length = length;
+				longest = &(plugin->*member);
+			}
+		}
+
+		return *longest;
+	}
+
+	void SetStringParameters(int widget) const
+	{
+		switch (widget) {
+			case WID_GO_SOCIAL_PLUGIN_TITLE:
+				/* For SetupSmallestSize, use the longest string we have. */
+				if (this->current_index < 0) {
+					SetDParamStr(0, GetWidestPlugin(&SocialIntegrationPlugin::name));
+					SetDParamStr(1, GetWidestPlugin(&SocialIntegrationPlugin::version));
+					break;
+				}
+
+				if (this->plugins[this->current_index]->name.empty()) {
+					SetDParam(0, STR_JUST_RAW_STRING);
+					SetDParamStr(1, this->plugins[this->current_index]->basepath);
+				} else {
+					SetDParam(0, STR_GAME_OPTIONS_SOCIAL_PLUGIN_TITLE);
+					SetDParamStr(1, this->plugins[this->current_index]->name);
+					SetDParamStr(2, this->plugins[this->current_index]->version);
+				}
+				break;
+
+			case WID_GO_SOCIAL_PLUGIN_PLATFORM:
+				/* For SetupSmallestSize, use the longest string we have. */
+				if (this->current_index < 0) {
+					SetDParamStr(0, GetWidestPlugin(&SocialIntegrationPlugin::social_platform));
+					break;
+				}
+
+				SetDParamStr(0, this->plugins[this->current_index]->social_platform);
+				break;
+
+			case WID_GO_SOCIAL_PLUGIN_STATE: {
+				/* For SetupSmallestSize, use the longest string we have. */
+				if (this->current_index < 0) {
+					SetDParam(0, STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_UNSUPPORTED_API);
```

Comment:
I did a thing. Mostly because I really didn't feel like making this `State` enum iteratable with `_BEGIN` and `_END` stuff. It makes it so ugly. And also because I don't like `STR_... + i` stuff :P It always makes my code fail when I change things :D

So I went for a different approach ... but not sure what people think. The compiler likes it equally, as in, with `-O3` there is no difference between this and a switch-case. So that is something.

Reply:
IMHO if you don't know what the longest string might be, and you don't want to test all possible combinations, it's better to design it around a wrapping mutli-line string instead. Especially if some translations might be quite long.

I guess the video driver description in the game options window is one example of where this is done.

Reply:
For context, the string in question is `Steam is not running`. And I mostly got myself in this position as it is not a regular window, but a composed one :P
Wrapping would be a bit odd here, even for translations, in how it would look. So currently it makes the window wider. But I am not really sure this is the right approach.

Maybe I should just not try to put things this neatly, and just put it below each other, give it a width, and let wrapping happen when it needs to happen.

Meh :P